proxy/coco/src/peer.rs

```diff
@@ -152,113 +145,52 @@ where
+            protocol_events,
+            subscriber,
+            control_receiver,
+        )
+        .map_err(Error::Join);
+
+        let protocol = async move {
+            loop {
+                match peer.bind().await {
+                    Ok(bound) => {
+                        match bound.listen_addrs() {
+                            Ok(listen_addrs) => {
+                                addrs_tx.send(listen_addrs).expect("subroutines is gone");
+                            },
+                            Err(e) => {
+                                return Err(Error::Io(e));
+                            },
+                        }
+
+                        if let Err(e) = net::protocol::accept(bound, disco.clone().discover()).await
+                        {
+                            log::error!("accept error: {}", e);
+                        }
+                    },
+                    Err(e) => {
+                        log::error!("bound error: {}", e);
+                        return Err(Error::Bootstrap(e));
+                    },
                 }
             }
-        });
-
-        Running {
-            protocol,
-            subroutines,
         }
-    }
-}
-
-/// Future returned by [`Peer::into_running`].
-#[must_use = "to the sig hup, don't stop, just drop"]
-pub struct Running {
-    /// Join and abort handles for the protocol run loop.
-    protocol: JoinHandle<Result<!, Error>>,
-    /// The [`Subroutines`] associated with this [`Peer`] instance.
-    subroutines: JoinHandle<Result<(), JoinError>>,
-}
+        .fuse();
+        futures::pin_mut!(protocol);
 
-impl Drop for Running {
-    fn drop(&mut self) {
-        log::trace!("`peer::Running` is being dropped");
-        self.protocol.abort();
-        self.subroutines.abort();
-    }
-}
-
-impl Future for Running {
-    type Output = Result<(), Error>;
-
-    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
-        let err = match self.protocol.poll_unpin(cx) {
-            Poll::Ready(val) => match val {
-                Err(e) => Some(Error::Join(e)),
-                Ok(res) => Some(into_err(res)),
-            },
-            Poll::Pending => None,
+        futures::select! {
+            res = protocol => res?,
+            res = subroutines => res?
         };
-
-        if let Some(err) = err {
-            log::trace!("run loop error: {:?}", err);
-            return Poll::Ready(Err(err));
-        }
-
-        match self.subroutines.poll_unpin(cx) {
-            Poll::Ready(val) => {
-                let val = match val {
-                    Err(e) | Ok(Err(e)) => Err(Error::Join(e)),
-                    Ok(Ok(())) => Ok(()),
-                };
-                Poll::Ready(val)
-            },
-            Poll::Pending => Poll::Pending,
-        }
-    }
-}
-
-fn into_err<T, E>(r: Result<T, E>) -> E
-where
-    T: Into<!>,
-{
-    match r {
-        Ok(x) => x.into(),
-        Err(e) => e,
+        Ok(())
```

Comment:
Can this ever be hit given that the output of the protocol future is `Result<!, Error>`?

Reply:
Yes. Like before, we return when the subroutine future is ready.